File: src/parallel_kinematics/pk.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

import sys
import os
import time
import math
import json
import threading

# Add huno_base/src to path to import wck and config
SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
HUNO_BASE_SRC = os.path.abspath(os.path.join(SCRIPT_DIR, '..', 'huno_base', 'src'))
sys.path.append(HUNO_BASE_SRC)

try:
    from wck import servo
    from config import DEFAULT_SERIAL_PORT
except ImportError:
    print("Error: Could not import wck or config from huno_base/src")
    print(f"Path searched: {HUNO_BASE_SRC}")
    sys.exit(1)

# File paths for configuration
JOINTS_FILE = os.path.join(HUNO_BASE_SRC, 'joints.json')
ZERO_FILE = os.path.join(HUNO_BASE_SRC, 'zero.json')

def load_json_file(filepath, default=None):
    if os.path.exists(filepath):
        try:
            with open(filepath, 'r') as f:
                return json.load(f)
        except Exception as e:
            print(f"Error loading {filepath}: {e}")
    return default

class ParallelKinematicsWalk:
    def __init__(self, port=DEFAULT_SERIAL_PORT, baud=115200):
        print(f"Connecting to servos on {port}...")
        try:
            self.wck = servo(port=port, baud=baud)
        except Exception as e:
            print(f"Failed to connect: {e}")
            sys.exit(1)

        self.joints_map = load_json_file(JOINTS_FILE, {})
        self.zero_map = load_json_file(ZERO_FILE, {})
        
        # Invert joints map: ID -> Name
        self.id_to_name = {v: k for k, v in self.joints_map.items()}
        
        # Parameters (Default from walk_open_loop_simple.py)
        self.f = 8.0            # Frequency
        self.robot_height = 0.5 # Knee bend offset (z)
        self.shift_y = 0.2      # Lateral shift (y)
        self.step_height = 0.4  # Step lift (z amplitude)
        self.step_length = 0.2  # Step length (x amplitude)
        self.arm_swing = 1.0    # Arm swing magnitude
        
        # Scaling factor from algorithm units to servo units
        self.scale = 30.0
        
        # Joint signs (to handle mounting directions)
        self.signs = {name: 1.0 for name in self.joints_map.keys()}
        # Initial guesses for HUNO sign symmetry
        # for name in self.signs:
        #     if "_r" in name:
        #         self.signs[name] = -1.0
        self.signs["j_ankle1_l"] = -1.0
        self.signs["j_ankle1_r"] = -1.0
        # self.signs["j_tibia_l"] = -1.0

        
        # self.signs["j_ankle2_r"] = -1.0
        # Specific overrides based on motion_controller.py and common biped logic
        # Pelvis Roll (0, 5), Ankle Roll (4, 9)
        # In HUNO, sometimes roll joints need inversion on both sides or same side.
        
        self.is_walking = False
        self.should_exit = False
        
        # Cache zero positions
        self.zeros = [127] * 16
        for name, joint_id in self.joints_map.items():
            if name in self.zero_map:
                self.zeros[joint_id] = self.zero_map[name]

        # Current target positions
        self.targets = list(self.zeros)

    def set_joint_val(self, name, val_offset):
        if name not in self.joints_map:
            return
        joint_id = self.joints_map[name]
        
        # Apply scaling and individual joint sign
        offset = int(val_offset * self.scale * self.signs[name])
        target = self.zeros[joint_id] + offset
        
        # Clamp to 1-254
        target = max(1, min(254, target))
        self.targets[joint_id] = target

    def left_leg(self, x, y, z):
        # Parallel Kinematics Algorithm
        self.set_joint_val("j_tibia_l",  z)          # Knee
        self.set_joint_val("j_thigh2_l", -z/2.0 + x) # Hip
        self.set_joint_val("j_ankle1_l", -z/2.0 - x) # Ankle Pitch
        
        self.set_joint_val("j_pelvis_l",  y)         # Hip Roll
        self.set_joint_val("j_ankle2_l", -y)         # Ankle Roll

    def right_leg(self, x, y, z):
        self.set_joint_val("j_tibia_r",  z)
        self.set_joint_val("j_thigh2_r", -z/2.0 + x)
        self.set_joint_val("j_ankle1_r", -z/2.0 - x)
        
        self.set_joint_val("j_pelvis_r",  y)
        self.set_joint_val("j_ankle2_r", -y)

    def update_arms(self, xLeft, xRight):
        self.set_joint_val("j_shoulder_l", self.arm_swing * xRight)
        self.set_joint_val("j_shoulder_r", self.arm_swing * xLeft)

    def run_walk_loop(self):
        print("Walk loop started.")
        while not self.should_exit:
            if self.is_walking:
                # Modulate frequency
                t = (time.time()) * self.f
                
                # lateral shift
                yLeftRight = math.sin(t) * self.shift_y
                
                # vertical lift (knee bend)
                zLeft  = (math.sin(t)           + 1.0) / 2.0 * self.step_height + self.robot_height
                zRight = (math.sin(t + math.pi) + 1.0) / 2.0 * self.step_height + self.robot_height

                # forward/backward swing
                xLeft  = math.cos(t)           * self.step_length
                xRight = math.cos(t + math.pi) * self.step_length
                
                self.left_leg(xLeft, yLeftRight, zLeft)
                self.right_leg(xRight, yLeftRight, zRight)
                self.update_arms(xLeft, xRight)
                
                # Synchronized move (all 16 possible joints)
                self.wck.posGroup(15, 4, self.targets)
            else:
                # Stand at zero
                self.wck.posGroup(15, 4, self.zeros)
                time.sleep(0.1)
                
            time.sleep(0.02) # ~50Hz

    def start(self):
        self.is_walking = True
        print("\n>>> Walking STARTED")

    def stop(self):
        self.is_walking = False
        print("\n>>> Walking STOPPED")

    def print_params(self):
        print("\n--- Current Parameters ---")
        print(f"  (f)  Frequency:    {self.f:<8.2f} (h)  Robot Height: {self.robot_height:.2f}")
        print(f"  (y)  Shift Y:      {self.shift_y:<8.2f} (sh) Step Height:  {self.step_height:.2f}")
        print(f"  (sl) Step Length:   {self.step_length:<8.2f} (as) Arm Swing:    {self.arm_swing:.2f}")
        print(f"  (sc) Servo Scale: {self.scale:<8.2f}")
        print("  Status: " + ("WALKING" if self.is_walking else "STOPPED"))

    def interactive_menu(self):
        self.print_params()
        print("\nCommands: start, stop, q (quit), params (show)")
        print("To change value: <param_code> <value> (e.g., 'f 10')")
        print("To toggle joint sign: sign <joint_name> (e.g., 'sign j_tibia_l')")
        
        while not self.should_exit:
            try:
                line = input("PK> ").strip().lower()
            except EOFError:
                break
                
            if not line: continue
            
            parts = line.split()
            cmd = parts[0]
            
            if cmd == 'q':
                self.should_exit = True
                self.stop()
            elif cmd == 'start':
                self.start()
            elif cmd == 'stop':
                self.stop()
            elif cmd == 'params':
                self.print_params()
            elif cmd == 'sign' and len(parts) > 1:
                name = parts[1]
                if name in self.signs:
                    self.signs[name] *= -1
                    print(f"Sign of {name} now {self.signs[name]}")
                else:
                    print(f"Unknown joint: {name}")
            elif len(parts) > 1:
                try:
                    val = float(parts[1])
                    if cmd == 'f': self.f = val
                    elif cmd == 'h': self.robot_height = val
                    elif cmd == 'y': self.shift_y = val
                    elif cmd == 'sh': self.step_height = val
                    elif cmd == 'sl': self.step_length = val
                    elif cmd == 'as': self.arm_swing = val
                    elif cmd == 'sc': self.scale = val
                    else: print(f"Unknown parameter: {cmd}")
                except ValueError:
                    print("Invalid value format")
            else:
                print("Unknown command. Type 'params' to see shortcuts.")


if __name__ == "__main__":
    pk = ParallelKinematicsWalk()
    
    # Run walk loop in a separate thread
    walk_thread = threading.Thread(target=pk.run_walk_loop)
    walk_thread.daemon = True
    walk_thread.start()
    
    try:
        pk.interactive_menu()
    except KeyboardInterrupt:
        pk.wck.posGroup(15, 4, pk.zeros)
        pk.should_exit = True
    
    print("Exiting...")
    pk.wck.close()

File: zmp/esp32_firmware/src/PK.cpp
#include "PK.h"
#include <math.h>

PK::PK(WCK* wck) : _wck(wck) {
    for(int i=0; i<NUM_JOINTS; i++) {
        targets[i] = zero_offsets[i];
        signs[i] = 1.0;
    }
    signs[J_ANKLE1_R] = -1.0;
    signs[J_TIBIA_L] = -1.0;
    signs[J_THIGH2_L] = -1.0;
}

void PK::begin() {
    _wck->posGroup(15, 4, zero_offsets);
}

void PK::startWalk() {
    is_walking = true;
}

void PK::stopWalk() {
    is_walking = false;
}

void PK::set_joint_val(int joint_id, float val_offset) {
    float offset = val_offset * scale * signs[joint_id];
    int target = zero_offsets[joint_id] + (int)offset;
    if (target < 1) target = 1;
    if (target > 254) target = 254;
    targets[joint_id] = target;
}

void PK::left_leg(float x, float y, float z) {
    set_joint_val(J_TIBIA_L, z);
    set_joint_val(J_THIGH2_L, -z/2.0 + x);
    set_joint_val(J_ANKLE1_L, -z/2.0 - x);
    set_joint_val(J_PELVIS_L, y);
    set_joint_val(J_ANKLE2_L, -y);
}

void PK::right_leg(float x, float y, float z) {
    set_joint_val(J_TIBIA_R, z);
    set_joint_val(J_THIGH2_R, -z/2.0 + x);
    set_joint_val(J_ANKLE1_R, -z/2.0 - x);
    set_joint_val(J_PELVIS_R, y);
    set_joint_val(J_ANKLE2_R, -y);
}

void PK::update_arms(float xLeft, float xRight) {
    set_joint_val(J_SHOULDER_L, arm_swing * xRight);
    set_joint_val(J_SHOULDER_R, arm_swing * xLeft);
}

void PK::printParams() {
    Serial.println("\n--- Current Parameters ---");
    Serial.print("  (f)  Frequency:    "); Serial.print(f);
    Serial.print("  (h)  Robot Height: "); Serial.println(robot_height);
    Serial.print("  (y)  Shift Y:      "); Serial.print(shift_y);
    Serial.print("  (sh) Step Height:  "); Serial.println(step_height);
    Serial.print("  (sl) Step Length:  "); Serial.print(step_length);
    Serial.print("  (ss) Side Step:    "); Serial.println(side_step);
    Serial.print("  (as) Arm Swing:    "); Serial.print(arm_swing);
    Serial.print("  (d)  Direction:    "); Serial.println(direction);
    Serial.print("  (sc) Servo Scale:  "); Serial.println(scale);
    Serial.print("  Status: "); Serial.println(is_walking ? "WALKING" : "STOPPED");
}

void PK::toggleSign(int joint_id) {
    if (joint_id >= 0 && joint_id < NUM_JOINTS) {
        signs[joint_id] *= -1.0;
        Serial.print("Sign of ID "); Serial.print(joint_id); Serial.print(" is now "); Serial.println(signs[joint_id]);
    }
}

void PK::interpolate_pose(uint8_t* start_pose, uint8_t* end_pose, int duration_ms, int torque) {
    int step_time_ms = 20;
    int steps = duration_ms / step_time_ms;
    if(steps < 1) steps = 1;
    
    for(int step = 1; step <= steps; step++) {
        uint8_t interpolated[NUM_JOINTS];
        float ratio = (float)step / steps;
        for(int i = 0; i < NUM_JOINTS; i++) {
            int val = start_pose[i] + (end_pose[i] - start_pose[i]) * ratio;
            if (val < 1) val = 1;
            if (val > 254) val = 254;
            interpolated[i] = val;
        }
        _wck->posGroup(15, torque, interpolated);
        delay(step_time_ms);
    }
    _wck->posGroup(15, torque, end_pose);
}

// --- GankenKun ZMP Preview Control Variables ---
// These gains should be pre-calculated in Python using scipy.linalg.solve_discrete_are
const float dt = 0.01; 
const float z_c = 0.27; // CoM Height
const float A_d[3][3] = {{1.0, dt, dt*dt/2.0}, {0.0, 1.0, dt}, {0.0, 0.0, 1.0}};
const float B_d[3] = {dt*dt*dt/6.0, dt*dt/2.0, dt};
const float C_d[3] = {1.0, 0.0, -z_c / 9.81};

// Example pre-calculated Gains (Mock values, MUST be recalculated for exact robot)
const float F_gain[4] = {-200.0, -50.0, -10.0, -1.0};
// -------------------------------------------

float com_x[3] = {0,0,0}; // CoM State X
float com_y[3] = {0,0,0}; // CoM State Y

void PK::update() {
    if (is_walking) {
        // 1. ZMP Preview Control: Calculate Control Input (u)
        // Here we would typically multiply the error and future ZMP references with F and f_gain.
        float u_x = 0.0; // Mock control input
        float u_y = 0.0; 

        // 2. Update CoM State using State Space Model (x = Ax + Bu)
        float next_com_x[3], next_com_y[3];
        for (int i=0; i<3; i++) {
            next_com_x[i] = A_d[i][0]*com_x[0] + A_d[i][1]*com_x[1] + A_d[i][2]*com_x[2] + B_d[i]*u_x;
            next_com_y[i] = A_d[i][0]*com_y[0] + A_d[i][1]*com_y[1] + A_d[i][2]*com_y[2] + B_d[i]*u_y;
        }
        for (int i=0; i<3; i++) { com_x[i] = next_com_x[i]; com_y[i] = next_com_y[i]; }

        // 3. Foot Trajectory Generation (Mock implementation)
        // Here we generate the X, Y, Z of the swinging foot based on time 't'.
        float t = (millis() / 1000.0) * direction * f;
        float zLeft = (sin(t) + 1.0) / 2.0 * step_height + robot_height;
        float zRight = (sin(t + PI) + 1.0) / 2.0 * step_height + robot_height;
        float xLeft = com_x[0] + cos(t) * step_length; // Relative to CoM
        float xRight = com_x[0] + cos(t + PI) * step_length;
        float yLeft = com_y[0] + shift_y;
        float yRight = com_y[0] - shift_y;

        // 4. Inverse Kinematics
        // Use Analytical IK instead of simple joint offsets
        left_leg(xLeft, yLeft, zLeft);
        right_leg(xRight, yRight, zRight);
        update_arms(xLeft, xRight);
        
        // 5. Send to Motors
        _wck->posGroup(15, 4, targets);
        was_walking = true;
        // No delay here! Controlled by FreeRTOS vTaskDelayUntil in main.cpp
    } else {
        if (was_walking) {
            interpolate_pose(targets, zero_offsets, 500);
            for(int i=0; i<NUM_JOINTS; i++) targets[i] = zero_offsets[i];
            was_walking = false;
        } else {
            // Keep zero pose active
            // _wck->posGroup(15, 4, zero_offsets);
        }
    }
}
